File: LeetCode-Python/CopyListwithRandomPointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head: return None
        # build copies
        node = head
        while node:
            copy = Node(node.val, node.next, node.random)
            node.next = copy
            node = copy.next
        
        # update next and random pointers to appropriate node
        node = head.next
        while node:
            if node.next:
                node.next = node.next.next # copy of the next node
            if node.random:
                node.random = node.random.next # copy of the random node
            node = node.next
        return head.next

File: LeetCode-Cpp/CopyListwithRandomPointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        
        if (!head) return nullptr;

        // Orijinal düğümler ve kopya düğümler arasındaki haritayı oluştur
        unordered_map<Node*, Node*> map;

        // İlk geçiş: Her düğüm için kopya oluştur ve haritaya ekle
        Node* node = head;
        while (node) {
            map[node] = new Node(node->val);
            node = node->next;
        }

        // İkinci geçiş: next ve random işaretçilerini ayarla
        node = head;
        while (node) {
            map[node]->next = map[node->next];
            map[node]->random = map[node->random];
            node = node->next;
        }

        // Kopyalanan listenin başını döndür
        return map[head];
    }
};
